Adaptation asides to be exported to the XML

StraNNiKK opened the pull request:
Continue working on xblock asides. Asides can be exported with the rest of courseware through the Export function within CMS.

P.S. This commit is continuation of the work that was started in
https://github.com/edx/edx-platform/pull/11562

edx-webhook commented:
Thanks for the pull request, @StraNNiKK! I've created [OSPR-1189](https://openedx.atlassian.net/browse/OSPR-1189) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

nedbat commented:
@cpennington More asides work!

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split_mongo_kvs.py:
```diff
@@ -148,7 +148,10 @@ def has(self, key):
 
-            b_type = key.block_scope_id.block_type
+            try:
+                b_type = key.block_scope_id.block_type
```
`block_scope_id` wouldn't have a `block_type` if the key is an `Aside`?

StraNNiKK replied:
@cpennington generally (for example for [StructuredTagsAside](https://github.com/edx/edx-platform/blob/master/cms/lib/xblock/tagging.py#L68)) `key.block_scope_id` is an object of `AsideUsageKeyV1` type. But there is one [dependency](https://github.com/edx/edx-platform/blob/master/requirements/edx/github.txt#L77) - it named `AcidAside`. I don't know why but in case of `AcidAside` object `block_scope_id` is always string:
![pycharm](https://cloud.githubusercontent.com/assets/1876893/13797247/282c629a-eb1d-11e5-950f-afed18a2e71f.jpeg)

cpennington replied:
Ah. That looks like a bug in the `split_mongo_kvs`. In particular, (as per your screenshot), the scope being used for this field is `Scope(user=UserScope.NONE, block=BlockScope.TYPE)`. In that scope, the `block_scope_id` should just be the block type (which it is, in this example). I think your best bet in this case would be to raise an error if the key scope isn't one of `Scope.content` or `Scope.settings` (as I think those are the only scopes that split-mongo handles right now).

cpennington replied:
In fact, you can use the same logic from line 120 (which looks at `self.VALID_SCOPES`). I think the existing code for `has` was being lenient and simply letting unknown scopes fall through, but it would probably be best `return False` early if the key doesn't have the right scope.

StraNNiKK replied:
Ok. I've changed this behaviour

cpennington commented on common/lib/xmodule/xmodule/xml_module.py:
```diff
@@ -428,6 +429,13 @@ def add_xml_to_node(self, node):
+        for aside in self.runtime.get_asides(self):
+            if aside.needs_serialization():
+                aside_node = etree.Element("unknown_root", nsmap=XML_NAMESPACES)
+                aside.has_children = False
```
This seems odd. `has_children` has typically been a class attribute on XBlocks. Perhaps we need it set at the top level of the XBlockAside base-class? Or perhaps the code that failed when you left this out should be using a `getattr` with a default?

StraNNiKK replied:
@cpennington 
`has_children` attribute is set in [ChildrenModelMetaclass](https://github.com/edx/XBlock/blob/master/xblock/mixins.py#L333) metaclass. `ChildrenModelMetaclass` is metaclass for [HierarchyMixin](https://github.com/edx/XBlock/blob/master/xblock/mixins.py#L355). `XBlock` inherits `HierarchyMixin`:
```
class XBlock(XmlSerializationMixin, HierarchyMixin, ScopedStorageMixin,
RuntimeServicesMixin, HandlersMixin, IndexInfoMixin, ViewsMixin, SharedBlockBase):
```
but `XBlockAside` isn't:
```
class XBlockAside(XmlSerializationMixin, ScopedStorageMixin, 
RuntimeServicesMixin, HandlersMixin, SharedBlockBase):
```
The `self.has_children` check is used in `XmlSerializationMixin.add_xml_to_node` function. It is the part of [xblock](https://github.com/edx/XBlock/blob/master/xblock/mixins.py#L487) project. So how should I correct code in this situation?

cpennington replied:
That's a bug in xblock that it would be great if you could fix. `HierarchyMixin` and `XmlSerializationMixin` are independent (as you gathered), so we shouldn't assume the former in the code for the latter.

StraNNiKK replied:
@cpennington I've created the related PR to fix this bug: https://github.com/edx/XBlock/pull/337

jbarciauskas commented:
Since edX/XBlock#337 is fixed, what else needs to happen here? @cpennington @StraNNiKK

StraNNiKK commented:
@jbarciauskas I've updated the PR and changed xblock version in the requirements from 0.4.7 to 0.4.8 (it was released 18 hours ago). Waiting for the response from @cpennington

StraNNiKK commented:
jenkins run python

StraNNiKK commented:
jenkins run js

StraNNiKK commented:
jenkins run js

StraNNiKK commented:
@cpennington could you please take a look

cpennington commented:
:+1:
